### PR TITLE
Added DSB and tests

### PR DESCRIPTION
**Pull request by hussenmi**
Added the DSB normalisation method and tests to work locally as well as in Docker. I also added a utils.py file and modified test_modules.py to work with the tests in Docker.

### CODE REVIEW

File: dockers/hto-adt-postprocess/run_docker_tests.sh
Comment (Tobiaspk) [inline dockers/hto-adt-postprocess/run_docker_tests.sh:1]: This uses. version 0.3.8. When you source `config.sh` first you can automatically use the most current version.

File: dockers/hto-adt-postprocess/run_tests.sh
Comment (Tobiaspk) [inline dockers/hto-adt-postprocess/run_tests.sh:1]: I'ld suggest to remove `black` and `ruff` from this manual testing script. In the next version we can add those to `pre-commit` to automatically run those tools. Same got for the corresponding `_venv.sh` script.

File: dockers/hto-adt-postprocess/src/demux_dsb.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        - dict: A dictionary containing the evaluation metrics for each HTO.
+    """
+    # Read in adata_filtered
+    adata_filtered = ad.read_h5ad(path_dsb_denoised_adata_dir)
+
+    # Check if the dsb_normalized is added in adata_filtered layers
+    if "dsb_normalized" in adata_filtered.layers:
+        df_umi_dsb_values = adata_filtered.layers['dsb_normalized']
+    else:
+        df_umi_dsb_values = adata_filtered.X
```
Comment (Tobiaspk) [inline dockers/hto-adt-postprocess/src/demux_dsb.py]: Better avoid to automatically default to `X`. You could add a `layer` parameter that defaults to `dsb_normalized`.

File: dockers/hto-adt-postprocess/src/demux_dsb.py
Comment (Tobiaspk) [inline dockers/hto-adt-postprocess/src/demux_dsb.py:1]: I looks like the function `evaluate_clustering` clusters the data a second time. This has two downfalls:
1. Initialisation is random and can lead to varying outputs
2. It's slower

I suggest to refactor `evaluate_metrics` to two separate functions that fit, predict and evaluate the clusters. Those functions are then called in the main loop of the `hto_demux_dsb` function. For example, create two functions as below:

- `run_gmm(data)`
- `run_kmeans(data)`

where each returns a dictionary like this:

```
{
"classification": <np-array>,
"metrics": <metrics-dict>,
}
```

File: dockers/hto-adt-postprocess/tests/test_demux_dsb.py
Comment (Tobiaspk) [inline dockers/hto-adt-postprocess/tests/test_demux_dsb.py:1]: Some more tests to consider:

- Inspect `gmm` and `kmeans` metrics:
  - What if the array is constant 0?
  - Does an array with clearly separated groups actually return the "highest" score and vice-versa?
- Moch data:
  - Instead of purely random, I suggest to create for example 3 groups with some signal to test that this signal is actually captured. For example

```
from sklearn.datasets import make_blobs

# generate random clusters
# return X and true y
mu = [(2, 1, 1), (1, 2, 1), (1, 1, 2)]
std = [.5, .5, .5]
X, y = make_blobs(n_samples=n, cluster_std=std, centers=muy, n_features=3, random_state=1)
```

**Review by Tobiaspk — COMMENTED** [review]
Added some comments. I suggest to avoid clustering multiple times in `evaluate_clustering` and adding one more test to demultiplexing.

File: dockers/hto-adt-postprocess/run_tests.sh
Comment (hussenmi) [reply dockers/hto-adt-postprocess/run_tests.sh:1]: Removed them for now

File: dockers/hto-adt-postprocess/run_docker_tests.sh
Comment (hussenmi) [reply dockers/hto-adt-postprocess/run_docker_tests.sh:1]: Changed to "latest" in `config.sh`

File: dockers/hto-adt-postprocess/src/demux_dsb.py (hunk 1 above)
Comment (hussenmi) [reply dockers/hto-adt-postprocess/src/demux_dsb.py]: actually, since we're always going to have the `dsb_normalized` layer, maybe this should be removed. that's by default. i've adjusted `dsb_algorithm.py` to always add the layer instead of adjusting the `X`

File: dockers/hto-adt-postprocess/src/demux_dsb.py
Comment (hussenmi) [reply dockers/hto-adt-postprocess/src/demux_dsb.py:1]: for the multiple clusterings, i had set the same random state, so the initial states were not completely random. but I just made a function called `cluster_and_evaluate` that works for both methods (means and gmm), and we now only do the clustering once. this returns data in the same way along the lines of what you mentioned

File: dockers/hto-adt-postprocess/tests/test_demux_dsb.py
Comment (hussenmi) [reply dockers/hto-adt-postprocess/tests/test_demux_dsb.py:1]: i also added a couple of more tests to see if the metrics are good indicators of when we have a well-separated clusters as well as when we have overlapping clusters

**Review by Tobiaspk — APPROVED** [review]
